File: week2/python/canny.py
# /bin/env python
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
#
# This code is made available to the students of
# the online course titled "Computer Vision for Faces"
# by Satya Mallick for personal non-commercial use.
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC.
#
# For licensing and other inquiries, please email
# spmallick@bigvisionllc.com

import cv2
import numpy as np
import argparse

parser = argparse.ArgumentParser(description='Apply canny filter to image')
parser.add_argument('file', type=str, nargs='?',
                    help='the image')

args = parser.parse_args()

imageFile = '../data/images/sample.jpg'
if args.file:
	imageFile = args.file

lowThreshold = 50
highThreshold = 100

maxThreshold = 1000

apertureSizes = [3, 5, 7]
maxapertureIndex = 2
apertureIndex = 0

blurAmount = 0
maxBlurAmount = 20

# Function for all trackbar calls
def applyCanny():
	# Blur the image before edge detection
	if(blurAmount > 0):
		blurredSrc = cv2.GaussianBlur(src, (2 * blurAmount + 1, 2 * blurAmount + 1), 0);
	else:
		blurredSrc = src.copy()

	# Canny requires aperture size to be odd
	apertureSize = apertureSizes[apertureIndex];

	# Apply canny to detect the images
	edges = cv2.Canny( blurredSrc, lowThreshold, highThreshold, apertureSize = apertureSize )

	# Display images
	cv2.imshow("Edges",edges)

# Function to update low threshold value
def updateLowThreshold( *args ):
	global lowThreshold
	lowThreshold = args[0]
	applyCanny()
	pass

# Function to update high threshold value
def updateHighThreshold( *args ):
	global highThreshold
	highThreshold = args[0]
	applyCanny()
	pass

# Function to update blur amount
def updateBlurAmount( *args ):
	global blurAmount
	blurAmount = args[0]
	applyCanny()
	pass

# Function to update aperture index
def updateApertureIndex( *args ):
	global apertureIndex
	apertureIndex = args[0]
	applyCanny()
	pass

# Read lena image
src = cv2.imread(imageFile, cv2.IMREAD_GRAYSCALE)

edges = src.copy()
# Display images
cv2.namedWindow("Edges", cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO)
cv2.resizeWindow('Edges', 800,800)
cv2.imshow("Edges", src)

# Trackbar to control the low threshold
cv2.createTrackbar( "Low Threshold", "Edges", lowThreshold, maxThreshold, updateLowThreshold)

# Trackbar to control the high threshold
cv2.createTrackbar( "High Threshold", "Edges", highThreshold, maxThreshold, updateHighThreshold)

# Trackbar to control the aperture size
cv2.createTrackbar( "aperture Size", "Edges", apertureIndex, maxapertureIndex, updateApertureIndex)

# Trackbar to control the blur
cv2.createTrackbar( "Blur", "Edges", blurAmount, maxBlurAmount, updateBlurAmount)


cv2.waitKey(0)
cv2.destroyAllWindows()

File: week2/cpp/canny.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "Computer Vision for Faces" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/

// Include OpenCV header files
#include <opencv2/opencv.hpp>

// Use cv and std namespaces
using namespace cv;
using namespace std;

// Variables for source and edge images
Mat src, edges;

// Variables for low and high thresholds
int lowThreshold = 50;
int highThreshold = 100;

// Max trackbar value
int const maxThreshold = 1000;

// Sobel aperture sizes for Canny edge detector
int apertureSizes [] = {3, 5, 7};
int maxapertureIndex = 2;
int apertureIndex = 0;

// Gaussian blur size
int blurAmount = 0;
int maxBlurAmount = 20;

// Function for trackbar call
void applyCanny(int ,void *)
{
  // Variable to store blurred image
  Mat blurredSrc;
  
  // Blur the image before edge detection
  if (blurAmount > 0 )
  {
    GaussianBlur(src, blurredSrc, Size( 2 * blurAmount + 1, 2 * blurAmount + 1), 0);
  }
  else
  {
    blurredSrc = src.clone();
  }
  
  // Canny requires aperture size to be odd
  int apertureSize = apertureSizes[apertureIndex];
  
  // Apply canny to get the edges
  Canny( blurredSrc, edges, lowThreshold, highThreshold, apertureSize );
  
  //Display images
  imshow("Edges",edges);
}

int main(void)
{
  // Read lena image
  src = imread("../data/images/sample.jpg", IMREAD_GRAYSCALE);
  
  //Display images
  imshow("Edges",src);
  
  // Create a window to display output.
  namedWindow("Edges",CV_WINDOW_AUTOSIZE);

  // Trackbar to control the low threshold
  createTrackbar( "Low Threshold", "Edges", &lowThreshold, maxThreshold, applyCanny);
  
  // Trackbar to control the high threshold
  createTrackbar( "High Threshold", "Edges", &highThreshold, maxThreshold, applyCanny);
  
  // Trackbar to control the aperture size
  createTrackbar( "aperture Size", "Edges", &apertureIndex, maxapertureIndex, applyCanny);
  
  // Trackbar to control the blur
  createTrackbar( "Blur", "Edges", &blurAmount, maxBlurAmount, applyCanny);

  waitKey(0);
  destroyAllWindows();

}
